python/repeated.c

```diff
@@ -298,6 +295,69 @@ static PyObject* PyUpb_RepeatedContainer_Subscript(PyObject* _self,
   }
 }
 
+static int PyUpb_RepeatedContainer_SetSubscript(
+    PyUpb_RepeatedContainer* self, upb_array* arr, const upb_fielddef* f,
+    Py_ssize_t idx, Py_ssize_t count, Py_ssize_t step, PyObject* value) {
+  upb_arena* arena = PyUpb_Arena_Get(self->arena);
+  if (upb_fielddef_issubmsg(f)) {
+    PyErr_SetString(PyExc_TypeError, "does not support assignment");
+    return -1;
+  }
+
+  if (step == 0) {
+    // Set single value.
+    upb_msgval msgval;
+    if (!PyUpb_PyToUpb(value, f, &msgval, arena)) return -1;
+    upb_array_set(arr, idx, msgval);
+    return 0;
+  }
+
+  // Set range.
+  PyObject* seq =
+      PySequence_Fast(value, "must assign iterable to extended slice");
+  if (!seq) return -1;
+  if (PySequence_Size(seq) != count) {
+    PyErr_Format(PyExc_ValueError,
+                  "attempt to assign sequence of size %zd to extended slice "
+                  "of size %zd",
+                  PySequence_Size(seq), count);
+    Py_DECREF(seq);
+    return -1;
+  }
+  for (Py_ssize_t i = 0; i < count; i++, idx += step) {
+    upb_msgval msgval;
+    PyObject* item = PySequence_GetItem(seq, i);
+    // XXX: if this fails we can leave the list partially mutated.
+    bool ok = PyUpb_PyToUpb(item, f, &msgval, arena);
+    Py_DECREF(item);
+    if (!ok) return -1;
+    upb_array_set(arr, idx, msgval);
+  }
+  return 0;
+}
+
+static int PyUpb_RepeatedContainer_DeleteSubscript(upb_array* arr,
+                                                   Py_ssize_t idx,
+                                                   Py_ssize_t count,
+                                                   Py_ssize_t step) {
+  // This is a delete, not a set.
+  // Normalize direction: deletion is order-independent.
+  if (step > 0) {
+    idx += step * (count - 1);
+    step = -step;
+  }
+  if (step == -1) {
+    // Contiguous range.
+    upb_array_delete(arr, idx - (count - 1), count);
+  } else {
+    // Stepped slice.
+    for (Py_ssize_t i = 0; i < count; i++, idx += step) {
+      upb_array_delete(arr, idx, 1);
```

Comment:
this will be quadratic because of repeated compactions.  I think we want to do something better here

Reply:
Done (no longer quadratic).